File: src/main/java/miretz/ycloud/services/MongoDBService.java
package miretz.ycloud.services;

import static com.mongodb.client.model.Filters.eq;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import javax.inject.Singleton;

import miretz.ycloud.services.utils.PasswordHashUtil;

import org.bson.Document;

import com.github.fakemongo.Fongo;
import com.google.inject.Inject;
import com.google.inject.name.Named;
import com.mongodb.MongoClient;
import com.mongodb.MongoCredential;
import com.mongodb.ServerAddress;
import com.mongodb.client.FindIterable;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;

@Singleton
public class MongoDBService implements DatabaseService {

	private static final String PROPERTY_FILES_DB_TABLE_NAME = "filesDbTableName";
	private static final String PROPERTY_USERS_DB_TABLE_NAME = "usersDbTableName";
	private static final String PROPERTY_ADMIN_USER = "adminUser";
	private static final String PROPERTY_ADMIN_PASS = "adminPass";
	protected static final String PROPERTY_DB_HOST = "dbHost";
	protected static final String PROPERTY_DB_PASS = "dbPass";
	protected static final String PROPERTY_DB_USER = "dbUser";
	protected static final String PROPERTY_USE_FAKE_MONGO = "useFakeMongo";
	protected static final String PROPERTY_DB_NAME = "dbName";

	protected MongoClient mongo;
	protected MongoDatabase database;

	protected String adminPass;
	protected String adminUser;
	protected String usersDbTableName;
	protected String filesDbTableName;

	@Inject
	public MongoDBService(@Named(PROPERTY_USE_FAKE_MONGO) String useFakeMongo, @Named(PROPERTY_DB_NAME) String dbName, @Named(PROPERTY_DB_HOST) String dbHost, @Named(PROPERTY_DB_PASS) String dbPass, @Named(PROPERTY_DB_USER) String dbUser, @Named(PROPERTY_ADMIN_USER) String adminUser,
			@Named(PROPERTY_ADMIN_PASS) String adminPass, @Named(PROPERTY_USERS_DB_TABLE_NAME) String usersDbTableName, @Named(PROPERTY_FILES_DB_TABLE_NAME) String filesDbTableName) {

		this.adminPass = adminPass;
		this.adminUser = adminUser;
		this.usersDbTableName = usersDbTableName;
		this.filesDbTableName = filesDbTableName;

		if (Boolean.parseBoolean(useFakeMongo)) {

			// fake mongo
			Fongo fongo = new Fongo("fake mongo server");
			database = fongo.getDatabase(dbName);

		} else {

			// Real mongo
			MongoCredential credential = MongoCredential.createMongoCRCredential(dbUser, dbName, dbPass.toCharArray());
			mongo = new MongoClient(new ServerAddress(dbHost), Arrays.asList(credential));
			database = mongo.getDatabase(dbName);

		}

		adminUserSetup();
		rootFolderSetup();
	}

	public void rootFolderSetup() {
		addDocument(new miretz.ycloud.models.Document("root", "root", "", new HashMap<String, String>(), miretz.ycloud.models.Document.TYPE_FOLDER));
	}

	public void adminUserSetup() {

		// drop all users
		// database.getCollection(USERS_TABLE).drop();

		// remove admin
		removeUser(adminUser);

		MongoCollection<Document> table = database.getCollection(usersDbTableName);

		Document document = new Document();
		document.put("username", adminUser);
		document.put("password", PasswordHashUtil.createHash(adminPass));
		document.put("createdDate", new Date());

		table.insertOne(document);
	}

	@Override
	public boolean checkUserPassword(String username, String password) {

		MongoCollection<Document> table = database.getCollection(usersDbTableName);
		FindIterable<Document> users = table.find(eq("username", username));

		for (Document user : users) {
			String pValue = user.getString("password");
			if (PasswordHashUtil.validatePassword(password, pValue)) {
				return true;
			}
		}

		return false;

	}

	@Override
	public void addUser(String username, String password) {
		String hash = PasswordHashUtil.createHash(password);
		if (hash != null) {
			MongoCollection<Document> table = database.getCollection(usersDbTableName);

			Document document = new Document();
			document.put("username", username);
			document.put("password", hash);
			document.put("createdDate", new Date());

			table.insertOne(document);
		}
	}

	@Override
	public void removeUser(String username) {
		MongoCollection<Document> table = database.getCollection(usersDbTableName);
		table.deleteOne(eq("username", username));
	}

	@Override
	public List<String> listUsernames() {
		MongoCollection<Document> table = database.getCollection(usersDbTableName);

		List<String> result = new ArrayList<String>();

		for (Document cur : table.find()) {
			result.add(cur.getString("username"));
		}

		return result;
	}

	/**
	 * DOCUMENT METHODS
	 */

	@Override
	public void deleteDocument(String contentId) {

		MongoCollection<Document> table = database.getCollection(filesDbTableName);

		// delete all children
		table.deleteMany(eq("parentId", contentId));

		// delete the document
		table.deleteOne(eq("contentId", contentId));

	}

	@Override
	public void addDocument(miretz.ycloud.models.Document documentToStore) {
		MongoCollection<Document> table = database.getCollection(filesDbTableName);

		Document document = new Document();
		document.append("contentId", documentToStore.getContentId());
		document.append("fileName", documentToStore.getFileName());
		document.append("parentId", documentToStore.getParentId());
		document.append("type", documentToStore.getType());

		Document metadata = new Document();
		for (Entry<String, String> entry : documentToStore.getMetadata().entrySet()) {
			metadata.append(entry.getKey(), entry.getValue());
		}
		document.append("metadata", metadata);
		document.append("createdDate", new Date());

		table.insertOne(document);

	}

	@Override
	public List<miretz.ycloud.models.Document> getAllDocuments() {
		MongoCollection<Document> table = database.getCollection(filesDbTableName);
		List<miretz.ycloud.models.Document> result = new ArrayList<>();

		for (Document document : table.find()) {
			result.add(getDocument(document));
		}

		return result;
	}

	@Override
	public List<miretz.ycloud.models.Document> getDescendants(String parentId) {
		MongoCollection<Document> table = database.getCollection(filesDbTableName);

		List<miretz.ycloud.models.Document> result = new ArrayList<>();

		FindIterable<Document> myDoc = table.find(eq("parentId", parentId));

		if (myDoc == null) {
			return result;
		}

		for (Document document : myDoc) {
			result.add(getDocument(document));
		}

		return result;

	}

	@Override
	public miretz.ycloud.models.Document getDocument(String contentId) {

		MongoCollection<Document> table = database.getCollection(filesDbTableName);

		Document document = table.find(eq("contentId", contentId)).first();

		if (document != null) {
			return getDocument(document);

		}

		return null;

	}

	private Map<String, String> getMetadataAsMap(Document document) {
		Map<String, String> metadataMap = new HashMap<>();
		for (Entry<String, Object> entry : document.get("metadata", Document.class).entrySet()) {
			metadataMap.put(entry.getKey(), (String) entry.getValue());
		}
		return metadataMap;
	}

	@Override
	public miretz.ycloud.models.Document findByMetadataValue(String key, String value) {
		MongoCollection<Document> table = database.getCollection(filesDbTableName);

		Document document = table.find(eq(key, value)).first();

		if (document != null) {
			return getDocument(document);
		}

		return null;
	}

	private miretz.ycloud.models.Document getDocument(Document document) {
		return new miretz.ycloud.models.Document(document.getString("contentId"), document.getString("fileName"), document.getString("parentId"), getMetadataAsMap(document), document.getString("type"));
	}

	@Override
	public void deleteAllDocuments() {
		MongoCollection<Document> table = database.getCollection(filesDbTableName);
		table.drop();
	}
}

File: src/main/java/miretz/ycloud/services/MongoDBService.kt
package miretz.ycloud.services

import com.github.fakemongo.Fongo
import com.google.inject.Inject
import com.google.inject.name.Named
import com.mongodb.MongoClient
import com.mongodb.MongoCredential
import com.mongodb.ServerAddress
import com.mongodb.client.MongoDatabase
import com.mongodb.client.model.Filters.eq
import miretz.ycloud.services.utils.PasswordHashUtil
import org.bson.Document
import java.util.*
import javax.inject.Singleton

@Singleton
class MongoDBService
@Inject
constructor(
        @Named(MongoDBService.PROPERTY_USE_FAKE_MONGO) useFakeMongo: String,
        @Named(MongoDBService.PROPERTY_DB_NAME) dbName: String,
        @Named(MongoDBService.PROPERTY_DB_HOST) dbHost: String,
        @Named(MongoDBService.PROPERTY_DB_PASS) dbPass: String,
        @Named(MongoDBService.PROPERTY_DB_USER) dbUser: String,

        @Named("adminUser") protected var adminUser: String,
        @Named("adminPass") protected var adminPass: String,
        @Named("usersDbTableName") protected var usersDbTableName: String,
        @Named("filesDbTableName") protected var filesDbTableName: String) : DatabaseService {

    companion object {
        const val PROPERTY_DB_HOST: String = "dbHost"
        const val PROPERTY_DB_PASS: String = "dbPass"
        const val PROPERTY_DB_USER: String = "dbUser"
        const val PROPERTY_USE_FAKE_MONGO: String = "useFakeMongo"
        const val PROPERTY_DB_NAME: String = "dbName"
    }

    protected var database: MongoDatabase

    init {

        if (java.lang.Boolean.parseBoolean(useFakeMongo)) {

            // fake mongo
            val fongo = Fongo("fake mongo server")
            database = fongo.getDatabase(dbName)

        } else {

            // Real mongo
            val credential = MongoCredential.createMongoCRCredential(dbUser, dbName, dbPass.toCharArray())
            val mongo = MongoClient(ServerAddress(dbHost), Arrays.asList(credential))
            database = mongo.getDatabase(dbName)

        }

        adminUserSetup()
        rootFolderSetup()
    }

    fun rootFolderSetup() {
        addDocument(miretz.ycloud.models.Document("root", "root", "", HashMap<String, String>(), miretz.ycloud.models.Document.TYPE_FOLDER))
    }

    fun adminUserSetup() {

        // drop all users
        // database.getCollection(USERS_TABLE).drop();

        // remove admin
        removeUser(adminUser)

        val table = database.getCollection(usersDbTableName)
        val document = Document().append("username", adminUser).append("password", PasswordHashUtil.createHash(adminPass)).append("createdDate", Date())

        table.insertOne(document)
    }

    override fun checkUserPassword(username: String, password: String): Boolean {

        val table = database.getCollection(usersDbTableName)
        val users = table.find(eq("username", username))

        for (user in users) {
            val pValue = user.getString("password")
            if (PasswordHashUtil.validatePassword(password, pValue)) {
                return true
            }
        }

        return false

    }

    override fun addUser(username: String, password: String) {
        val hash = PasswordHashUtil.createHash(password) ?: IllegalStateException("Cannot hash password")
        val table = database.getCollection(usersDbTableName)
        val document = Document().append("username", username).append("password", hash).append("createdDate", Date())
        table.insertOne(document)
    }

    override fun removeUser(username: String) {
        val table = database.getCollection(usersDbTableName)
        table.deleteOne(eq("username", username))
    }

    override fun listUsernames(): List<String> {
        val table = database.getCollection(usersDbTableName)
        return table.find().map { it.getString("username") }.toList()
    }

    /**
     * DOCUMENT METHODS
     */

    override fun deleteDocument(contentId: String) {

        val table = database.getCollection(filesDbTableName)

        // delete all children
        table.deleteMany(eq("parentId", contentId))

        // delete the document
        table.deleteOne(eq("contentId", contentId))

    }

    override fun addDocument(documentToStore: miretz.ycloud.models.Document) {
        val table = database.getCollection(filesDbTableName)

        val document = Document()
        document.append("contentId", documentToStore.contentId)
        document.append("fileName", documentToStore.fileName)
        document.append("parentId", documentToStore.parentId)
        document.append("type", documentToStore.type)

        val metadata = Document()
        for (entry in documentToStore.metadata.entries) {
            metadata.append(entry.key, entry.value)
        }
        document.append("metadata", metadata)
        document.append("createdDate", Date())

        table.insertOne(document)

    }

    override fun getAllDocuments(): List<miretz.ycloud.models.Document> {
        val table = database.getCollection(filesDbTableName)
        val result : List<miretz.ycloud.models.Document> = table.find().map { it -> getDocument(it) }.toList()
        return result
    }

    override fun getDescendants(parentId: String): List<miretz.ycloud.models.Document> {
        val table = database.getCollection(filesDbTableName)
        val myDoc = table.find(eq<String>("parentId", parentId)) ?: return ArrayList()
        return myDoc.map { it -> getDocument(it) }.toList()
    }


    override fun findDocument(contentId: String): miretz.ycloud.models.Document {
        val table = database.getCollection(filesDbTableName)
        val document = table.find(eq("contentId", contentId)).first() ?: throw IllegalStateException("Document not found")
        return getDocument(document)
    }

    private fun getMetadataAsMap(document: Document): Map<String, String> {
        val metadataMap = HashMap<String, String>()
        for (entry in document.get("metadata", Document().javaClass).entries) {
            metadataMap.put(entry.key, entry.value as String)
        }
        return metadataMap
    }

    override fun findByMetadataValue(key: String, value: String): miretz.ycloud.models.Document {
        val table = database.getCollection(filesDbTableName)
        val document = table.find(eq(key, value)).first() ?: throw IllegalStateException("Document not found")
        return getDocument(document)
    }

    private fun getDocument(document: Document): miretz.ycloud.models.Document {
        return miretz.ycloud.models.Document(document.getString("contentId"), document.getString("fileName"), document.getString("parentId"), getMetadataAsMap(document), document.getString("type"))
    }

    override fun deleteAllDocuments() {
        val table = database.getCollection(filesDbTableName)
        table.drop()
    }

}
